File: go/anagram/anagram.go
package anagram

import "strings"

const testVersion = 1

func Detect(s string, dic []string) []string {
	result := []string{}
	for _, word := range dic {
		if isAnagram(strings.ToLower(s), strings.ToLower(word)) {
			result = append(result, strings.ToLower(word))
		}
	}
	return result
}

func isAnagram(s string, d string) bool {
	return (s != d) && (sortString(s) == sortString(d))
}

func sortString(s string) string {
	r := []rune(s)
	for i := 0; i < len(r); i++ {
		min := i
		for j := i; j < len(r); j++ {
			if r[min] > r[j] {
				min = j
			}
		}
		r[i], r[min] = r[min], r[i]
	}
	return string(r)
}

File: cpp/anagram/anagram.cpp
#include "anagram.h"

namespace anagram {
  anagram::anagram(string in){
    targetWord = in;
  }
  vector<string> anagram::matches(vector<string> const& words){
    vector<string> results;
    for(auto const& word : words){
      if (isAnagram(targetWord, word)){
        results.push_back(word);
      }
    }
    return results;
  }

  bool anagram::isAnagram(string const& left, string const& right){
    string sortedLeft = left;
    string sortedRight = right;
    boost::to_lower(sortedLeft);
    boost::to_lower(sortedRight);
    if (sortedLeft == sortedRight)
      return false;
    sort(sortedLeft.begin(), sortedLeft.end());
    sort(sortedRight.begin(), sortedRight.end());
    if (sortedLeft == sortedRight)
      return true;
    return false;
  }
}
